### PR TITLE
MOTS-150: Validation on saving Incharges.

### PR DESCRIPTION
**Pull request by dserkowski**

### CODE REVIEW

File: src/main/java/org/motechproject/mots/constants/MotsConstants.java [hunk 1]
```diff
@@ -3,4 +3,6 @@
 public final class MotsConstants {
   public static final String SIMPLE_DATE_FORMAT = "yyyy-MM-dd";
 
+  public static final String UUID_PATTERN =
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/constants/MotsConstants.java]: if it's not used anywhere else move it to UuidValidator

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java [hunk 2]
```diff
@@ -2,19 +2,26 @@
...
 public class InchargeDto {
 
   @Getter
   @Setter
+  @Uuid(message = ValidationMessages.INVALID_INCHARGE_ID)
   private String id;
 
   @Getter
   @Setter
+  @NotEmpty(message = ValidationMessages.EMPTY_FIST_NAME)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/dto/InchargeDto.java]: I think we could use NotBlank instead

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java [hunk 3]
```diff
@@ -2,19 +2,26 @@
...
   private String firstName;
 
   @Getter
   @Setter
+  @NotEmpty(message = ValidationMessages.EMPTY_SECOND_NAME)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/dto/InchargeDto.java]: as above

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java [hunk 4]
```diff
@@ -23,22 +30,30 @@
 
   @Getter
   @Setter
+  @NotEmpty(message = ValidationMessages.EMPTY_PHONE_NUMBER)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/dto/InchargeDto.java]: as above

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java [hunk 5]
```diff
@@ -23,22 +30,30 @@
...
 
   @Getter
   @Setter
+  @Email(message = ValidationMessages.INVALID_EMAIL)
   private String email;
 
   @Getter
   @Setter
+  @NotEmpty
+  @Uuid(message = ValidationMessages.INVALID_DISTRICT_ID)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/dto/InchargeDto.java]: it is not required when inchange is saved, only the facilityId is

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java [hunk 6]
```diff
@@ -23,22 +30,30 @@
...
   private String districtId;
 
   @Getter
   @Setter
+  @NotEmpty
+  @Uuid(message = ValidationMessages.INVALID_CHIEFDOM_ID)
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/dto/InchargeDto.java]: as above

File: src/main/java/org/motechproject/mots/constants/ValidationMessages.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package org.motechproject.mots.constants;
+
+public final class ValidationMessages {
+
+  public static final String EMPTY_CHW_ID = "CHW Id cannot be empty";
+  public static final String EMPTY_FIST_NAME = "First name cannot be empty";
```
Comment (kkaczmarczyk) [inline src/main/java/org/motechproject/mots/constants/ValidationMessages.java]: EMPTY_FIST_NAME -> EMPTY_FIRST_NAME

File: src/main/java/org/motechproject/mots/web/InchargeController.java [hunk 8]
```diff
@@ -59,9 +66,10 @@ public InchargeDto getIncharge(@PathVariable("id") UUID id) {
   @RequestMapping(value = "/incharge", method = RequestMethod.POST)
   @ResponseStatus(HttpStatus.CREATED)
   @ResponseBody
-  public InchargeDto createIncharge(@RequestBody InchargeDto inchargeDto) {
+  public InchargeDto createIncharge(@RequestBody @Valid InchargeDto inchargeDto,
+      BindingResult bindingResult) {
     Incharge incharge = inchargeMapper.fromDto(inchargeDto);
-
+    validateIncharge(incharge, bindingResult);
```
[the remark this answers is not in the document]
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/web/InchargeController.java]: Ok, I will create new BindingResult.

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java (hunk 5 above)
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/dto/InchargeDto.java]: Ok, I don't notice that. I will remove UUID and notEmpty checks

File: src/main/java/org/motechproject/mots/constants/MotsConstants.java (hunk 1 above)
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/constants/MotsConstants.java]: ok

File: src/main/java/org/motechproject/mots/dto/InchargeDto.java (hunk 2 above)
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/dto/InchargeDto.java]: Yes, I think it is a good idea

File: src/main/java/org/motechproject/mots/web/InchargeController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pmuchowski) [reply src/main/java/org/motechproject/mots/web/InchargeController.java]: I think you could use dto instead, the only disadvantage is that you will need to parse the id's or use toString on uuid's

File: src/main/java/org/motechproject/mots/constants/ValidationMessages.java (hunk 7 above)
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/constants/ValidationMessages.java]: Ok

File: src/main/java/org/motechproject/mots/web/InchargeController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dserkowski) [reply src/main/java/org/motechproject/mots/web/InchargeController.java]: Ok

**Comment by pmuchowski** [comment]
@dserkowski resolve the conflicts

File: src/main/java/org/motechproject/mots/validate/constraintvalidators/FacilityExistenceValidator.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  @Override
+  public boolean isValid(String facilityId, ConstraintValidatorContext context) {
+    boolean isValid = true;
+    if (StringUtils.isNotEmpty(facilityId) && ValidationUtils.isValidUuidString(facilityId)) {
+      UUID facility = UUID.fromString(facilityId);
+      if (!facilityRepository.exists(facility)) {
+        isValid = false;
+      }
+    }
+    return isValid;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/constraintvalidators/FacilityExistenceValidator.java]: use return true here

File: src/main/java/org/motechproject/mots/validate/constraintvalidators/FacilityExistenceValidator.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  @Autowired
+  private FacilityRepository facilityRepository;
+
+  @Override
+  public boolean isValid(String facilityId, ConstraintValidatorContext context) {
+    boolean isValid = true;
+    if (StringUtils.isNotEmpty(facilityId) && ValidationUtils.isValidUuidString(facilityId)) {
+      UUID facility = UUID.fromString(facilityId);
+      if (!facilityRepository.exists(facility)) {
+        isValid = false;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/constraintvalidators/FacilityExistenceValidator.java]: use return false here

File: src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoFacilityUniquenessValidator.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    ConstraintValidator<FacilityUniqueness, InchargeDto> {
+
+  private static final String FACILITY_ID = "facilityId";
+
+  @Autowired
+  private InchargeRepository inchargeRepository;
+
+  @Override
+  public boolean isValid(InchargeDto inchargeDto, ConstraintValidatorContext context) {
+    boolean isValid = true;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoFacilityUniquenessValidator.java]: as above

File: src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoFacilityUniquenessValidator.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    if (StringUtils.isNotEmpty(inchargeDto.getId())
+        && StringUtils.isNotEmpty(inchargeDto.getFacilityId())
+        && ValidationUtils.isValidUuidString(inchargeDto.getFacilityId())) {
+      UUID facilityId = UUID.fromString(inchargeDto.getFacilityId());
+      Optional<Incharge> existingIncharge = inchargeRepository
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoFacilityUniquenessValidator.java]: I think it's possible to findByFacilityId, you can use it instead

File: src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoPhoneNumberUniquenessValidator.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    ConstraintValidator<PhoneNumberUniqueness, InchargeDto> {
+
+  private static final String PHONE_NUMBER = "phoneNumber";
+
+  @Autowired
+  private InchargeRepository inchargeRepository;
+
+  @Override
+  public boolean isValid(InchargeDto inchargeDto, ConstraintValidatorContext context) {
+    boolean isValid = true;
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/validate/constraintvalidators/incharge/InchargeDtoPhoneNumberUniquenessValidator.java]: you can return values directly and remove this variable

File: src/main/java/org/motechproject/mots/web/InchargeController.java [hunk 14]
```diff
@@ -75,12 +79,16 @@ public InchargeDto createIncharge(@RequestBody InchargeDto inchargeDto) {
...
+      @RequestBody @Valid InchargeDto inchargeDto, BindingResult bindingResult) {
+    checkBindingResult(bindingResult);
     Incharge existingIncharge = inchargeService.getIncharge(id);
-
     inchargeMapper.updateFromDto(inchargeDto, existingIncharge);
-
     return inchargeMapper.toDto(inchargeService.saveIncharge(existingIncharge));
   }
 
+  private void checkBindingResult(BindingResult bindingResult) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/InchargeController.java]: move it to BaseController, so it can be used in other controllers too

**Pull request merged**